### PR TITLE
[SPARK-31854][SQL] Invoke in MapElementsExec should not propagate null

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This PR intends to fix a bug of `Dataset.map` below when the whole-stage codegen enabled;
```
scala> val ds = Seq(1.asInstanceOf[Integer], null.asInstanceOf[Integer]).toDS()

scala> sql("SET spark.sql.codegen.wholeStage=true")

scala> ds.map(v=>(v,v)).explain
== Physical Plan ==
*(1) SerializeFromObject [assertnotnull(input[0, scala.Tuple2, true])._1.intValue AS _1#69, assertnotnull(input[0, scala.Tuple2, true])._2.intValue AS _2#70]
+- *(1) MapElements <function1>, obj#68: scala.Tuple2
   +- *(1) DeserializeToObject staticinvoke(class java.lang.Integer, ObjectType(class java.lang.Integer), valueOf, value#1, true, false), obj#67: java.lang.Integer
      +- LocalTableScan [value#1]

// `AssertNotNull` in `SerializeFromObject` will fail;
scala> ds.map(v => (v, v)).show()
java.lang.NullPointerException: Null value appeared in non-nullable fails:
top level Product input object
If the schema is inferred from a Scala tuple/case class, or a Java bean, please try to use scala.Option[_] or other nullable types (e.g. java.lang.Integer instead of int/scala.Int).

// When the whole-stage codegen disabled, the query works well;
scala> sql("SET spark.sql.codegen.wholeStage=false")
scala> ds.map(v=>(v,v)).show()
+----+----+
|  _1|  _2|
+----+----+
|   1|   1|
|null|null|
+----+----+
```
A root cause is that `Invoke` used in `MapElementsExec` propagates input null, and then [AssertNotNull](https://github.com/apache/spark/blob/1b780f364bfbb46944fe805a024bb6c32f5d2dde/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala#L253-L255) in `SerializeFromObject` fails because a top-level row becomes null. So, `MapElementsExec` should not return `null` but `(null, null)`.

NOTE: the generated code of the query above in the current master;
```
/* 033 */   private void mapelements_doConsume_0(java.lang.Integer mapelements_expr_0_0, boolean mapelements_exprIsNull_0_0) throws java.io.IOException {
/* 034 */     boolean mapelements_isNull_1 = true;
/* 035 */     scala.Tuple2 mapelements_value_1 = null;
/* 036 */     if (!false) {
/* 037 */       mapelements_resultIsNull_0 = false;
/* 038 */
/* 039 */       if (!mapelements_resultIsNull_0) {
/* 040 */         mapelements_resultIsNull_0 = mapelements_exprIsNull_0_0;
/* 041 */         mapelements_mutableStateArray_0[0] = mapelements_expr_0_0;
/* 042 */       }
/* 043 */
/* 044 */       mapelements_isNull_1 = mapelements_resultIsNull_0;
/* 045 */       if (!mapelements_isNull_1) {
/* 046 */         Object mapelements_funcResult_0 = null;
/* 047 */         mapelements_funcResult_0 = ((scala.Function1) references[1] /* literal */).apply(mapelements_mutableStateArray_0[0]);
/* 048 */
/* 049 */         if (mapelements_funcResult_0 != null) {
/* 050 */           mapelements_value_1 = (scala.Tuple2) mapelements_funcResult_0;
/* 051 */         } else {
/* 052 */           mapelements_isNull_1 = true;
/* 053 */         }
/* 054 */
/* 055 */       }
/* 056 */     }
/* 057 */
/* 058 */     serializefromobject_doConsume_0(mapelements_value_1, mapelements_isNull_1);
/* 059 */
/* 060 */   }
```

The generated code w/ this fix;
```
/* 032 */   private void mapelements_doConsume_0(java.lang.Integer mapelements_expr_0_0, boolean mapelements_exprIsNull_0_0) throws java.io.IOException {
/* 033 */     boolean mapelements_isNull_1 = true;
/* 034 */     scala.Tuple2 mapelements_value_1 = null;
/* 035 */     if (!false) {
/* 036 */       mapelements_mutableStateArray_0[0] = mapelements_expr_0_0;
/* 037 */
/* 038 */       mapelements_isNull_1 = false;
/* 039 */       if (!mapelements_isNull_1) {
/* 040 */         Object mapelements_funcResult_0 = null;
/* 041 */         mapelements_funcResult_0 = ((scala.Function1) references[1] /* literal */).apply(mapelements_mutableStateArray_0[0]);
/* 042 */
/* 043 */         if (mapelements_funcResult_0 != null) {
/* 044 */           mapelements_value_1 = (scala.Tuple2) mapelements_funcResult_0;
/* 045 */           mapelements_isNull_1 = false;
/* 046 */         } else {
/* 047 */           mapelements_isNull_1 = true;
/* 048 */         }
/* 049 */
/* 050 */       }
/* 051 */     }
/* 052 */
/* 053 */     serializefromobject_doConsume_0(mapelements_value_1, mapelements_isNull_1);
/* 054 */
/* 055 */   }
```

### Why are the changes needed?

Bugfix.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 1]
```diff
@@ -383,15 +383,26 @@ case class Invoke(
...
+        if (!${ev.isNull}) {
+          $evaluate
+        }
+      """
+    } else {
+      s"""
+        ${ev.isNull} = false;
+        $evaluate
+      """
+    }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:398]: Do we need this? When `propagateNull` is false, in `prepareArguments` `resultIsNull` will be set to false literal.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:398]: Ur, I missed that and you're right. I'll revert this. Thanks!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala [hunk 2]
```diff
@@ -202,6 +203,20 @@ class ObjectExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
       Invoke(funcSubObj, "binOp", DoubleType, inputSum), 0.75, InternalRow.apply(1, 0.25))
   }
 
+  test("SPARK-31854 Invoke in MapElementsExec should not propagate null") {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala:206]: nit: `SPARK-31854` -> `SPARK-31854:`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 3]
```diff
@@ -366,11 +366,15 @@ case class Invoke(
       // If the function can return null, we do an extra check to make sure our null bit is still
       // set correctly.
       val assignResult = if (!returnNullable) {
-        s"${ev.value} = (${CodeGenerator.boxedType(javaType)}) $funcResult;"
+        s"""
+          ${ev.value} = (${CodeGenerator.boxedType(javaType)}) $funcResult;
+          ${ev.isNull} = false;
```
Comment (TJX2014) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:371]: Seems we can remove ` ${ev.isNull} = false;`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 4]
```diff
@@ -366,11 +366,15 @@ case class Invoke(
...
+         """
       } else {
         s"""
           if ($funcResult != null) {
             ${ev.value} = (${CodeGenerator.boxedType(javaType)}) $funcResult;
+            ${ev.isNull} = false;
```
Comment (TJX2014) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:377]: Same as this

**Comment by maropu** [comment]
cc: @cloud-fan @viirya

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 5]
```diff
@@ -276,12 +276,12 @@ case class MapElementsExec(
...
 
   override def doConsume(ctx: CodegenContext, input: Seq[ExprCode], row: ExprCode): String = {
-    val (funcClass, methodName) = func match {
+    val (funcClass, funcName) = func match {
       case m: MapFunction[_, _] => classOf[MapFunction[_, _]] -> "call"
       case _ => FunctionUtils.getFunctionOneName(outputObjectType, child.output(0).dataType)
     }
     val funcObj = Literal.create(func, ObjectType(funcClass))
-    val callFunc = Invoke(funcObj, methodName, outputObjectType, child.output)
+    val callFunc = Invoke(funcObj, funcName, outputObjectType, child.output, propagateNull = false)
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala:284]: I changed the name `methodName`->`funcName` so as to put the code in a single line. plz let me know if I need to revert this.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 6]
```diff
@@ -1916,6 +1916,17 @@ class DatasetSuite extends QueryTest
     assert(df1.semanticHash !== df3.semanticHash)
     assert(df3.semanticHash === df4.semanticHash)
   }
+
+  test("SPARK-31854: Invoke in MapElementsExec should not propagate null") {
+    spark.conf.set("spark.sql.codegen.wholeStage", false)
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:1921]: why need to set this?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala [hunk 7]
```diff
@@ -202,6 +203,20 @@ class ObjectExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+    val inputInt = Seq(BoundReference(0, ObjectType(classOf[Any]), true))
+    val outputType = ObjectType(classOf[(Any, Any)])
+    val inputRow = InternalRow.fromSeq(Seq(null.asInstanceOf[java.lang.Integer]))
+    val createExpr = (propagateNull: Boolean) => {
+      Invoke(funcObj, "mapFunc", outputType, inputInt, propagateNull)
+    }
+    checkObjectExprEvaluation(createExpr(true), null, inputRow)
+    checkObjectExprEvaluation(createExpr(false), (null, null), inputRow)
+  }
+
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala:219]: I think this test is already passed before this change, right?

**Review by viirya — COMMENTED** [review]
Looks okay as I think we should not short circuit for null inputs (i.e. `propagateNull` = true) in `MapElementsExec`. The given function isn't necessarily returning null in all cases.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala (hunk 6 above)
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:1921]: Ur, I forgot to remove this...

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala (hunk 7 above)
Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ObjectExpressionsSuite.scala:219]: Ah, yes.

**Review by viirya — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala (hunk 5 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala:284]: It looks okay for that purpose.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by maropu** [comment]
Thanks, all!

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan . This is a bug since Apache Spark 2.0.x. Could you backport this to `branch-2.4` please, too?

**Comment by cloud-fan** [comment]
There are several conflicts in the test suite when I backport to 3.0, I believe there are more conflicts in 2.4. @maropu can you send a new PR for 2.4? thanks!

**Comment by dongjoon-hyun** [comment]
Oh, got it. Thanks.

**Comment by HyukjinKwon** [comment]
+1 for backporting. LGTM too

**Comment by maropu** [comment]
Sure, I will.

**Pull request closed without merging**